#2142. Change static analysis member invocation tests to be extension type tests

sgrekhov opened the pull request:
Move and rename these tests in the next PR

eernstg commented on LanguageFeatures/Inline-classes/static_analysis_member_invocation_A05_t03.dart:
```diff
@@ -2,52 +2,48 @@
-  final int id;
-  IC1(this.id);
+extension types ET1(int id) {
   IC1.n1(this.id, [String s = ""]);
   IC1.n2(this.id, {String s = ""});
   IC1.n3(this.id, {required String s});
```
```suggestion
  ET1.n1(this.id, [String s = ""]);
  ET1.n2(this.id, {String s = ""});
  ET1.n3(this.id, {required String s});
```

eernstg commented on LanguageFeatures/Inline-classes/static_analysis_member_invocation_A06_t01.dart:
```diff
@@ -2,35 +2,29 @@
-  final int id;
-  V(this.id);
-}
+extension type ET(int id) implements I {}
```
Again, `implements I` is a compile-time error because it doesn't satisfy `int <: I`. So we need to do something like `implements num`, and then we should be able to do `e.ceil()` where `e` has static type `ET`, but not `e.isEven`.

eernstg commented on LanguageFeatures/Inline-classes/static_analysis_member_invocation_A06_t02.dart:
```diff
@@ -2,29 +2,27 @@
-inline class V implements V1 {
-  V();
-}
+extension type ET(int id) implements I {}
```
Same issue: `implements I` is an error, we'd need something like `implements Comparable<num>` (just to try something new ;-), and then we should be able to do `e.compareTo(1.5)` where `e` has static type `ET`.

eernstg replied:
Alternatively, we could check things like these ones:

```dart
class I {
  int i = 0;
}

class J extends I {
  int j = 1;
}

extension type E(J rep) implements I {
  int get jOfE => rep.j;
}

void main() {
  var e = E(J());
  e.rep; // OK.
  e.i; // OK.
  e.j; // Compile-time error.
  e.jOfE; // OK.
}
```

eernstg commented on LanguageFeatures/Inline-classes/static_analysis_member_invocation_A07_t01.dart:
```diff
@@ -0,0 +1,49 @@
+  int foo<T1, T2 extends num>() => 42;
+}
+
+extension type IC2<T extends num>(T id) {
```
```suggestion
extension type ET2<T extends num>(T id) {
```

eernstg commented on LanguageFeatures/Inline-classes/static_analysis_member_invocation_A07_t01.dart:
```diff
@@ -0,0 +1,49 @@
+  T foo<T1, T2 extends T>() => id;
+}
+
+extension type IC3<T>(T id) {
```
```suggestion
extension type ET3<T>(T id) {
```

eernstg reviewed (commented):
Looks good! There are some typos, and there are a couple of cases where `implements T` is used in a situation where the given choice of `T` makes it a compile-time error.

sgrekhov commented:
Ups. I thought that we are allowed to implement any class in case of extension types. Updated tests, fixed typos. Thank you for noticing! Please rereview

sgrekhov commented:
Friendly ping

eernstg commented on LanguageFeatures/Inline-classes/static_analysis_member_invocation_A06_t03.dart:
```diff
@@ -0,0 +1,66 @@
+  var et3 = ET3(J());
+  et3.rep;
+  et3.i;
+  et3.j;
```
Delete this line?

eernstg commented on LanguageFeatures/Inline-classes/static_analysis_member_invocation_A06_t03.dart:
```diff
@@ -0,0 +1,66 @@
+///
+/// @description Checks that if an extension type `ET` has a superinterface with
+/// a member `m` then this member is also presents in `ET`v but members of its
+/// representation type
```
```suggestion
/// representation type are not.
```

eernstg reviewed (approved):
LGTM
